src/deluge/gui/views/session_view.cpp

```diff
@@ -4609,3 +4638,157 @@ Output* SessionView::getOutputFromPad(int32_t x, int32_t y) {
+	int32_t trackIndex = gridTrackIndexFromTrack(track, maxTrack);
+	x = gridXFromTrack(trackIndex);
+	y = gridYFromSection(clip->section);
+}
+
+// stop pulsing selected clip
+// called when exiting session view
+void SessionView::gridStopSelectedClipPulsing() {
+	uiTimerManager.unsetTimer(TimerName::SELECTED_CLIP_PULSE);
+	gridSelectedClipPulsing = false;
+	selectedClipForPulsing = nullptr;
+	gridResetSelectedClipPulseProgress();
+}
+
+// reset blend position for pulse
+void SessionView::gridResetSelectedClipPulseProgress() {
+	progress = kMinProgress;
+	blendDirection = 1;
+	gridSelectedClipRenderedColour = RGB::monochrome(0);
+}
+
+// render pulse for selected clip
+void SessionView::gridSelectClipForPulsing(Clip* clip) {
+	// save the selected clip so that we don't override the pulsed colour when re-rendering grid
+	gridResetSelectedClipPulseProgress();
+	selectedClipForPulsing = clip;
+}
+
+// check if we should stop pulsing
+bool SessionView::gridCheckForPulseStop() {
+	// don't render pulse if we're in another view
+	if (getCurrentUI() != this) {
+		return true;
+	}
+	// don't render pulse if we're in row view
+	else if (currentSong->sessionLayout == SessionLayoutType::SessionLayoutTypeRows) {
+		return true;
+	}
+	// don't render pulse if we're transiting from session view to another view
+	else if (currentUIMode == UI_MODE_EXPLODE_ANIMATION) {
+		return true;
+	}
+	return false;
+}
+
+static const uint32_t pulseUIModes[] = {UI_MODE_CLIP_PRESSED_IN_SONG_VIEW, UI_MODE_CREATING_CLIP, 0};
+
+void SessionView::gridPulseSelectedClip() {
+	if (gridSelectedClipPulsing) {
+		// check if we should stop pulsing
+		if (gridCheckForPulseStop()) {
+			return gridStopSelectedClipPulsing();
+		}
+
+		// is UI mode valid for pulsing?
+		if (isUIModeWithinRange(pulseUIModes)) {
+			Clip* currentClip = getCurrentClip();
+			Output* currentOutput = nullptr;
+			if (currentClip) {
+				currentOutput = currentClip->output;
+			}
+
+			// don't render pulse if you're holding record or the clip is armed and flashing
+			// don't render pulse if the track color has not been set yet
+			if (currentClip && currentOutput->colour && !viewingRecordArmingActive
+			    && !(view.clipArmFlashOn && currentClip->armState != ArmState::OFF)) {
+				// get the X, Y coordinates of the current clip
+				uint32_t selectedClipX;
+				uint32_t selectedClipY;
+				gridCoordsFromClip(currentClip, selectedClipX, selectedClipY);
+				// don't render pulse if clip is scrolled out of view (coords = - 1)
+				if (selectedClipX != -1 && selectedClipY != -1) {
+					// pulse if:
+					// A) you're not holding a pad
+					// or
+					// B) you're holding a pad that's different from the selected clip
+					if (currentUIMode == UI_MODE_NONE
+					    || (gridFirstPressedX != selectedClipX || gridFirstPressedY != selectedClipY)) {
+
+						// save the selected clip so that we don't override the pulsed colour when re-rendering grid
+						selectedClipForPulsing = currentClip;
```

Comment:
Why do these need to be different? If we're only ever pulsing the selected clip then it would make more sense to just check the same variable